File: src/app/service/company-request.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { CompanyRequest } from '../models/company-request.model';

@Injectable({
  providedIn: 'root'
})
export class CompanyRequestService {
  private baseUrl = 'http://localhost:8080/api/company-requests';

  constructor(private http: HttpClient) {}

  submitRequest(request: CompanyRequest, userId: number): Observable<CompanyRequest> {
    return this.http.post<CompanyRequest>(`${this.baseUrl}/submit/${userId}`, request);
  }

  getAllRequests(): Observable<CompanyRequest[]> {
    return this.http.get<CompanyRequest[]>(`${this.baseUrl}/all`);
  }

  approveRequest(requestId: number): Observable<CompanyRequest> {
    return this.http.post<CompanyRequest>(`${this.baseUrl}/approve/${requestId}`, {});
  }

  rejectRequest(requestId: number): Observable<void> {
    return this.http.post<void>(`${this.baseUrl}/reject/${requestId}`, {});
  }

  getApprovedCompaniesByService(service: string): Observable<CompanyRequest[]> {
    return this.http.get<CompanyRequest[]>(`${this.baseUrl}/approved/${service}`);
  }

  removeCompany(companyId: number): Observable<void> {
    return this.http.post<void>(`${this.baseUrl}/remove/${companyId}`, {});
  }
}

File: demo/src/main/java/com/example/demo/service/CompanyRequestService.java
package com.example.demo.service;

import com.example.demo.model.CompanyRequest;
import com.example.demo.model.User;
import com.example.demo.repository.CompanyRequestRepository;
import com.example.demo.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class CompanyRequestService {

    @Autowired
    private CompanyRequestRepository companyRequestRepository;

    @Autowired
    private UserRepository userRepository;

    public List<CompanyRequest> getApprovedCompaniesByService(String service) {
        return companyRequestRepository.findByServicesAndApproved(service, true);
    }

    public CompanyRequest submitRequest(CompanyRequest request, Long userId) {
        User user = userRepository.findById(userId).orElseThrow(() -> new RuntimeException("User not found"));
        request.setUser(user);
        return companyRequestRepository.save(request);
    }

    public List<CompanyRequest> getAllRequests() {
        return companyRequestRepository.findAll();
    }

    public CompanyRequest approveRequest(Long requestId) {
        CompanyRequest request = companyRequestRepository.findById(requestId).orElseThrow(() -> new RuntimeException("Request not found"));
        request.setApproved(true);
        User user = request.getUser();
        user.setRole("COMPANY");
        userRepository.save(user); // Save the updated user with the new role
        return companyRequestRepository.save(request);
    }

    public void rejectRequest(Long requestId) {
        companyRequestRepository.deleteById(requestId);
    }

    public void removeCompany(Long companyId) {
        CompanyRequest company = companyRequestRepository.findById(companyId).orElseThrow(() -> new RuntimeException("Company not found"));
        User user = company.getUser();
        user.setRole("USER");
        userRepository.save(user); // Save the updated user with the new role
        companyRequestRepository.deleteById(companyId); // Remove the company request
    }
}
